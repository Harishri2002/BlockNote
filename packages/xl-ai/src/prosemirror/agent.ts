import { BlockNoteEditor } from "@blocknote/core";
import { Transaction } from "prosemirror-state";
import { Mapping, ReplaceStep, Step } from "prosemirror-transform";

/**
 * Takes an array of ReplaceSteps and applies them as a human-like agent.
 * Every step is split into 3 phases:
 * - select the text to be replaced (1 transaction per ReplaceStep)
 * - replace the text with the first character of the replacement (if any) (1 transaction per ReplaceStep)
 * - insert the replacement character by character (strlen-1 transactions per ReplaceStep)
 *
 * All these phases are dispatched to the `dispatch` function as separate transactions.
 */
export async function applyStepsAsAgent(
  editor: BlockNoteEditor,
  steps: Step[],
  dispatch: (
    tr: Transaction,
    type: "select" | "replace" | "insert"
  ) => Promise<void>
) {
  // Create a mapping to track position changes across all steps
  const stepMapping = new Mapping();

  for (const step of steps) {
    if (!(step instanceof ReplaceStep)) {
      throw new Error("Step is not a ReplaceStep");
    }

    if (step.slice.openStart > 0 || step.slice.openEnd > 0) {
      throw new Error("Slice has openStart or openEnd > 0");
    }

    // Map the step positions through all previous mappings
    const mappedFrom = stepMapping.map(step.from);
    const mappedTo = stepMapping.map(step.to);

    // 1. Select text to be removed/replaced
    const selectTr = editor.prosemirrorState.tr.setMeta("aiAgent", {
      selection: {
        anchor: mappedFrom,
        head: mappedTo,
      },
    });
    await dispatch(selectTr, "select");

    const replacement =
      step.slice.size === 0 ? step.slice.content : step.slice.content.cut(0, 1);

    let replaceEnd = mappedTo;

    // TODO: option to replace instead of deletion / insertion
    const replaceTr = editor.prosemirrorState.tr
      .addMark(mappedFrom, replaceEnd, editor.pmSchema.mark("deletion", {}))
      .insert(replaceEnd, replacement)
      .addMark(
        replaceEnd,
        replaceEnd + replacement.size,
        editor.pmSchema.mark("insertion", {})
      )
      .setMeta("aiAgent", {
        selection: {
          anchor: replaceEnd + 1,
          head: replaceEnd + 1,
        },
      });

    await dispatch(replaceTr, "replace");

    stepMapping.appendMapping(replaceTr.mapping);
    replaceEnd = replaceTr.mapping.map(replaceEnd);

    // 3. Insert remaining characters one by one
    for (let i = 1; i < step.slice.size; i++) {
      const replacement = step.slice.content.cut(i, i + 1);
      const insertTr = editor.prosemirrorState.tr
        .insert(replaceEnd, replacement)
        .addMark(
          replaceEnd,
          replaceEnd + replacement.size,
          editor.pmSchema.mark("insertion", {})
        )
        .setMeta("aiAgent", {
          selection: {
            anchor: replaceEnd + 1,
            head: replaceEnd + 1,
          },
        });

      await dispatch(insertTr, "insert");
      stepMapping.appendMapping(insertTr.mapping);
      replaceEnd = insertTr.mapping.map(replaceEnd);
    }
  }
}
