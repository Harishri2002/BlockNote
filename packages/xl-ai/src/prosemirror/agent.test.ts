import { BlockNoteEditor, getBlockInfo, getNodeById } from "@blocknote/core";
import { Fragment, Slice } from "prosemirror-model";
import { Transaction } from "prosemirror-state";
import { ReplaceStep } from "prosemirror-transform";
import { describe, expect, it, vi } from "vitest";
import { applyStepsAsAgent } from "./agent.js";

describe("applyStepsAsAgent", () => {
  // Helper function to create a test editor with a simple paragraph
  function createTestEditor() {
    return BlockNoteEditor.create({
      initialContent: [
        {
          id: "1",
          type: "paragraph",
          content: "Hello, world!",
        },
      ],
    });
  }

  it("should apply a simple replace step", async () => {
    const editor = createTestEditor();

    // Get the position of the content in the paragraph
    const blockPos = getNodeById("1", editor.prosemirrorState.doc)!;
    const block = getBlockInfo(blockPos);
    if (!block.isBlockContainer) {
      throw new Error("Block is not a container");
    }

    const contentStart = block.blockContent.beforePos;

    // Create a ReplaceStep that replaces "Hello" with "Hi"
    const from = contentStart + 1; // +1 to skip the initial position
    const to = contentStart + 6; // "Hello" is 5 characters

    // Create a slice using the editor's schema
    const fragment = Fragment.from(editor.pmSchema.text("Hi"));

    // const slice = fragment.slice(0, fragment.content.size);

    const step = new ReplaceStep(from, to, new Slice(fragment, 0, 0));

    // Mock dispatch function
    const dispatch = vi
      .fn()
      .mockImplementation(async (tr: Transaction, type) => {
        editor.dispatch(tr);
        return (
          type + ": " + JSON.stringify(editor.prosemirrorState.doc.toJSON())
        );
      });

    // Apply the step
    await applyStepsAsAgent(editor, [step], dispatch);

    // Verify dispatch was called with the correct transactions
    expect(dispatch).toHaveBeenCalledTimes(3); // select, replace, insert

    expect(
      dispatch.mock.settledResults.map((r) => r.value).join("\n")
    ).toMatchSnapshot();
  });

  it("should handle multiple steps", async () => {
    const editor = createTestEditor();

    // Get the position of the content in the paragraph
    const blockPos = getNodeById("1", editor.prosemirrorState.doc)!;
    const block = getBlockInfo(blockPos);
    if (!block.isBlockContainer) {
      throw new Error("Block is not a container");
    }

    const contentStart = block.blockContent.beforePos;

    // Create two ReplaceSteps
    // 1. Replace "Hello" with "Hi"
    const fragment1 = Fragment.from(editor.pmSchema.text("Hi"));
    const step1 = new ReplaceStep(
      contentStart + 1,
      contentStart + 6,
      new Slice(fragment1, 0, 0)
    );

    // 2. Replace "world" with "there"
    const fragment2 = Fragment.from(editor.pmSchema.text("there"));
    const step2 = new ReplaceStep(
      contentStart + 8,
      contentStart + 13,
      new Slice(fragment2, 0, 0)
    );

    // Mock dispatch function
    const dispatch = vi
      .fn()
      .mockImplementation(async (tr: Transaction, type) => {
        editor.dispatch(tr);
        return (
          type + ": " + JSON.stringify(editor.prosemirrorState.doc.toJSON())
        );
      });

    // Apply the steps
    await applyStepsAsAgent(editor, [step1, step2], dispatch);

    // Verify dispatch was called for each step
    // For each step: select, replace, and potentially multiple inserts
    expect(dispatch).toHaveBeenCalledTimes(9);

    expect(
      dispatch.mock.settledResults.map((r) => r.value).join("\n")
    ).toMatchSnapshot();
  });

  it("should throw an error for non-ReplaceSteps", async () => {
    const editor = createTestEditor();

    // Create a non-ReplaceStep (we'll just mock it)
    const nonReplaceStep = { from: 0, to: 5 } as any;

    // Mock dispatch function
    const dispatch = vi.fn().mockImplementation(async () => Promise.resolve());

    // Expect the function to throw an error
    await expect(
      applyStepsAsAgent(editor, [nonReplaceStep], dispatch)
    ).rejects.toThrow("Step is not a ReplaceStep");
  });

  it("should throw an error for slices with openStart or openEnd > 0", async () => {
    const editor = createTestEditor();

    // Get the position of the content in the paragraph
    const blockPos = getNodeById("1", editor.prosemirrorState.doc)!;
    const block = getBlockInfo(blockPos);
    if (!block.isBlockContainer) {
      throw new Error("Block is not a container");
    }

    const contentStart = block.blockContent.beforePos;

    // Create a custom slice with openEnd > 0 (mocking the behavior)
    const step = new ReplaceStep(contentStart + 1, contentStart + 6, {
      content: editor.pmSchema.text("Hi"),
      openStart: 0,
      openEnd: 1,
      size: 2,
    } as any);

    // Mock dispatch function
    const dispatch = vi.fn().mockImplementation(async () => Promise.resolve());

    // Expect the function to throw an error
    await expect(applyStepsAsAgent(editor, [step], dispatch)).rejects.toThrow(
      "Slice has openStart or openEnd > 0"
    );
  });
});
