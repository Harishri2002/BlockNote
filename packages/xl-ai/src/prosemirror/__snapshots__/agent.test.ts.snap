// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`applyStepsAsAgent > should apply a simple replace step 1`] = `
"select: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","text":"Hello, world!"}]}]}]}]}
replace: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","marks":[{"type":"deletion"}],"text":"Hello"},{"type":"text","marks":[{"type":"insertion"}],"text":"H"},{"type":"text","text":", world!"}]}]}]}]}
insert: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","marks":[{"type":"deletion"}],"text":"Hello"},{"type":"text","marks":[{"type":"insertion"}],"text":"Hi"},{"type":"text","text":", world!"}]}]}]}]}"
`;

exports[`applyStepsAsAgent > should handle multiple steps 1`] = `
"select: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","text":"Hello, world!"}]}]}]}]}
replace: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","marks":[{"type":"deletion"}],"text":"Hello"},{"type":"text","marks":[{"type":"insertion"}],"text":"H"},{"type":"text","text":", world!"}]}]}]}]}
insert: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","marks":[{"type":"deletion"}],"text":"Hello"},{"type":"text","marks":[{"type":"insertion"}],"text":"Hi"},{"type":"text","text":", world!"}]}]}]}]}
select: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","marks":[{"type":"deletion"}],"text":"Hello"},{"type":"text","marks":[{"type":"insertion"}],"text":"Hi"},{"type":"text","text":", world!"}]}]}]}]}
replace: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","marks":[{"type":"deletion"}],"text":"Hello"},{"type":"text","marks":[{"type":"insertion"}],"text":"Hi"},{"type":"text","text":", "},{"type":"text","marks":[{"type":"deletion"}],"text":"world"},{"type":"text","marks":[{"type":"insertion"}],"text":"t"},{"type":"text","text":"!"}]}]}]}]}
insert: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","marks":[{"type":"deletion"}],"text":"Hello"},{"type":"text","marks":[{"type":"insertion"}],"text":"Hi"},{"type":"text","text":", "},{"type":"text","marks":[{"type":"deletion"}],"text":"world"},{"type":"text","marks":[{"type":"insertion"}],"text":"th"},{"type":"text","text":"!"}]}]}]}]}
insert: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","marks":[{"type":"deletion"}],"text":"Hello"},{"type":"text","marks":[{"type":"insertion"}],"text":"Hi"},{"type":"text","text":", "},{"type":"text","marks":[{"type":"deletion"}],"text":"world"},{"type":"text","marks":[{"type":"insertion"}],"text":"the"},{"type":"text","text":"!"}]}]}]}]}
insert: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","marks":[{"type":"deletion"}],"text":"Hello"},{"type":"text","marks":[{"type":"insertion"}],"text":"Hi"},{"type":"text","text":", "},{"type":"text","marks":[{"type":"deletion"}],"text":"world"},{"type":"text","marks":[{"type":"insertion"}],"text":"ther"},{"type":"text","text":"!"}]}]}]}]}
insert: {"type":"doc","content":[{"type":"blockGroup","content":[{"type":"blockContainer","attrs":{"id":"1","textColor":"default","backgroundColor":"default"},"content":[{"type":"paragraph","attrs":{"textAlignment":"left"},"content":[{"type":"text","marks":[{"type":"deletion"}],"text":"Hello"},{"type":"text","marks":[{"type":"insertion"}],"text":"Hi"},{"type":"text","text":", "},{"type":"text","marks":[{"type":"deletion"}],"text":"world"},{"type":"text","marks":[{"type":"insertion"}],"text":"there"},{"type":"text","text":"!"}]}]}]}]}"
`;
